File: cpp/maxareaofisland.cpp
class Solution {
public:
    int maxAreaOfIsland(vector<vector<int>>& grid) {
        if (grid.empty()) return 0;
        for (int i = 0; i < grid.size(); i++) {
            for (int j = 0; j < grid[0].size(); j++) {
                if (grid[i][j] == 1) {
                    int count = 0;
                    dfs(grid, i, j, count);
                }
            }
        }
        return maxNum;
    }
    void dfs(vector<vector<int>>& grid, int i, int j, int& count) {
        if (i < 0 or j < 0 or i == grid.size() or j == grid[0].size() or grid[i][j] == 0) {
            return;
        }
        grid[i][j] = 0;
        count++;
        maxNum = max(maxNum, count);
        dfs(grid, i-1, j, count);
        dfs(grid, i+1, j, count);
        dfs(grid, i, j-1, count);
        dfs(grid, i, j+1, count);
    }
    int maxNum = 0;
};

File: java/MaxAreaOfIsland.java
class Solution {
    
    int curMax;
    
    public int maxAreaOfIsland(int[][] grid) {
        int maxArea = 0;
        for (int i = 0; i < grid.length; i++) {
            for (int j = 0; j < grid[0].length; j++) {
                if (grid[i][j] != 0) {
                    curMax = 0;
                    dfs(grid, i, j);
                    System.out.println(curMax);
                    maxArea = Math.max(maxArea, curMax);
                }
            }
        }
        
        return maxArea;
    }
    
    private void dfs(int[][] grid, int i, int j) {
        
        if (i < 0 || i == grid.length || j < 0 || j == grid[i].length || grid[i][j] == 0) {
            return;
        }
        
        grid[i][j] = 0;
        curMax += 1;
        
        dfs(grid, i + 1, j);
        dfs(grid, i - 1, j);
        dfs(grid, i, j + 1);
        dfs(grid, i, j - 1);
        
    }
    
}
